### ISSUE TITLE
Variance of the reverse process explodes

### ISSUE THREAD

**Issue by jamesheald** [openai/improved-diffusion#105]
I have implemented a diffusion probabilistic model for the first time using a cosine noise schedule, and I am finding that the variance/standard deviation of samples drawn from the reverse process (whether the reverse process is untrained, or trained via Algorithm 1 of https://proceedings.neurips.cc/paper/2020/file/4c5bcfec8584af0d967f1ab10179ca4b-Paper.pdf) grows with the number of diffusion steps, producing samples that are well out of the distribution of the training data (which is of order 1 magnitude). My model does not seem to improve with training (samples from the 'trained' model look like noise, despite the loss coming down somewhat), and I feel as though this variance explosion, which seems pathological, is the cause. How I should think about/address this? 

In the improved-diffusion paper, there is a line that says 'We assume that image data consists of integers in {0, 1, ... , 255} scaled linearly to [-1, 1]. This ensures that the neural network reverse process operates on consistently scaled inputs starting from the standard normal prior p(x_T).' How are the constantly scaled inputs ensured during reverse diffusion?

I have given a MWE below, where I have replaced the noise predicted by a neural network with actual noise (this replacement isn't important, as the variance explosion happens if the noise is set to 0).

```
def cosine_beta_schedule(T, s = 0.008):
    """
    cosine schedule as proposed in https://proceedings.mlr.press/v139/nichol21a/nichol21a.pdf
    """
    t = np.linspace(0, T, T + 1)
    alphas_cumprod = np.cos((t / T + s) / (1 + s) * np.pi / 2) ** 2
    alphas_cumprod = alphas_cumprod / alphas_cumprod[0]
    betas = 1 - (alphas_cumprod[1:] / alphas_cumprod[:-1])
    betas_clipped = np.clip(betas, a_min = 0, a_max = 0.999)
    return betas_clipped

import numpy as np

T = 1000 # number of diffusion steps
betas = cosine_beta_schedule(T)
alphas = 1. - betas
alphas_cumprod = np.cumprod(alphas)
alphas_cumprod_prev = np.hstack([1., alphas_cumprod[:-1]])
sqrt_one_minus_alphas_cumprod = np.sqrt(1. - alphas_cumprod)
posterior_variance = np.clip(betas * (1. - alphas_cumprod_prev) / (1. - alphas_cumprod), a_min = 1e-20, a_max = 1)

# reverse process sampling loop (Algorithm 2 of https://proceedings.neurips.cc/paper/2020/file/4c5bcfec8584af0d967f1ab10179ca4b-Paper.pdf)
x_t = np.random.randn(10000) 
for t in np.arange(T)[::-1]:
        noise_pred = np.random.randn(10000) # THIS IS A SURROGATE FOR A NEURAL NETWORK THAT PREDICTS THE NOISE ADDED TO DATA
        posterior_mean = 1 / np.sqrt(alphas[t]) * (x_t - betas[t] / sqrt_one_minus_alphas_cumprod[t] * noise_pred)
        posterior_variance = betas[t] * (1. - alphas_cumprod_prev[t]) / (1. - alphas_cumprod[t])
        x_t = posterior_mean + np.sqrt(posterior_variance) * np.random.randn(10000)

print('standard deviation of', np.std(x_t))
```

**Comment by AxelCole** [comment]
Has anyone found a solution to this variance explosion when using a cosine schedule ? Thank you beforehand for any help !

**Comment by jamesheald** [comment]
You need to clip samples during reverse diffusion between (-1, 1), otherwise all hell breaks loose (see this post https://github.com/hojonathanho/diffusion/issues/5). It's a hack that is swept under the rug but seems essential for good performance in my experience.
